gazelle_haskell_modules/rule_generation.go

```diff
@@ -265,21 +267,60 @@ func concatRuleInfos(xs [][]*RuleInfo) []*RuleInfo {
 	return ys
 }
 
-// Collects the source files referenced in the given expression
-func srcsFromRule(pkgRoot string, expr build.Expr) ([]string, error) {
-	srcs, err := getSources(expr)
-	if err != nil {
-		return nil, err
+// Find either
+// 0. The files listed in the srcs attribute of a rule
+// 1. All the files (recursively) in directories specified in a "gazelle_haskell_modules:srcs:"
+// If the "srcs" attribute is present, 1. is ignored.
```

Comment:
Is this the behaviour we want? It seems like for a user that knows what they're doing, this is better. However, it might be confusing for someone trying out the project for the first time.

Reply:
If this is the desired behaviour, we should perhaps delete the comment containing the autodiscover directive, and perhaps mention that it didn't fire due to the presence of `srcs`.

Reply:
Won't the `haskell_module` rule fail though? I guess it won't even get to running if nothing depends on it. Is there some reason to keep it around if the underlying file is gone?

Reply:
> Won't the haskell_module rule fail though? 

If nothing depends on it, it won't be used during the build. Invoking gazelle with the fix command will delete rules with no enclosing library.

That said, IIRC the `modules` attribute is going to be preserved, which will point to the deleted files. This is a fine question to consider now. Should we delete modules from the `modules` attribute? Should we do this only when there is an autodiscovery directive?

Reply:
Answering myself after thinking it some more:

> Should we delete modules from the modules attribute? 

I think this is what the user might prefer. But if we don't do it, it must be tolerable for the user to remove the files manually, and possibly the implementation stays simpler.

> Should we do this only when there is an autodiscovery directive?

@aherrmann said to me that this is the sort of thing that the fix command would be expected to do. But I also think we can have the user remove the deleted modules manually from the configuration for now.

If it turns out later that editing the configuration after removals is too tedious, we can revise these choices.

Regarding the meaning of function here, I would blend the contents of the `srcs` attribute and the autodiscovered sources (unless there is some good reason to ignore the `srcs` attribute).

Lastly, enabling autodiscovery should produce an error if the rule is marked as `# gazelle_haskell_modules:keep` or `# keep`.

Reply:
We now combine them together.

Reply:
The part of erroring when both a keep and an autodetect directive are present will require a bit of a bigger change, since right now we don't even get to finding the autodetect directive if a keep one is present. I think that someone adding the keep directive would already expect the autodetect one to do nothing, so it doesn't seem like it's that important for an error to be produced in this case. 

Do you still want to have it done?

Reply:
> Do you still want to have it done?

No, there's no need to do it if it looks complicated.